### PR TITLE
Washing Machine - Coolification Expansion

### PR DESCRIPTION
**Pull request by Nakalon**
## About The Pull Request

Washing Machines have a cool thing going on for them, the recoloring of items.

However, in my view, a couple of things are missing to make it actually interesting and usable.

The aim of this PR is to expand upon this system, fix and tweak what needs fixing.

In the future I will expand on this system by taking clothes and converting them to GAGs for the purpose of dyeing them in the Washing Machine.

Things like Laser Tag armour, or scarves.

My idea is, you can Dye an item properly on the washing machine using the colours available (crayons and stamps), If you don't want any of those colours you can take your item, Dye it white with a sheet of paper or White Crayon and then spraycan the item in whatever colour you want.
Spraycanning clothing doesn't look nearly as good as Dyeing it in the washing machine but it does look passible on white colours, allowing players to Ghetto Style their clothes to their taste.

### What this PR does:

Expands the pool of items that can be dyed to include:
- Suits
- Hats
- Berets
- Clown only dyes for Banana Peels, Bananas and Practice Laser Guns.

Reworks the Denied Stamp "REDCOAT" dye into "DENIED" and adds sprites for the results of that dye on:
- Jumpsuits
- Jumpskirts
- Hats
- Gloves
- Shoes

(These are the ERROR items. I used to show people what would happen if we dyed a suit with the denied stamp, it resulted in the no-sprite cursed invisible suit with the words "SUIT" spelled out. This is meant to kind of give that bug vibe [keeping it thematically "naked] without making it an actual bug).

Added the SECURITY dye, used to dye items to make them look like security clothing.

Adds a couple of other sprites including:
- White Beret
- Rainbow Beret
- Mime Beret (The invisible beret)
- The HONK Tactical Beret

Adds sounds to the Washing Machines.

Fixes various bugs including:
- Jumpskirts not being Dyeable
- Shoes not being Dyeable
- Pet Gibbing producing a bugged Pet
- Cleaning not working

Adds new features such as:
- Washing an item that was Dyed now returns it to normal.
- Accessories now keep their colour when attached to an article of clothing that was spray-canned.
- Colour sources now survive the washing process and are no longer expended. (Stamps and Crayons)
- Adds SoP books and Paper as a colour source that can only be used once.
- Now you can colour all items of the same type, all gloves, all jumpskirts, all underclothing, etc.

### You might find it interesting that weapons and banana peels are listed here.

I like to throw some bones to roles that generally don't get new content, this is what I have in plan for clowns.

I want to give clowns a reason to use this system in their machinations, namely:

Having Practice Laser Guns that LOOK like bananas and banana peels that look like wet floor signs for example.

I am very interested in exploring this concept OUTSIDE of clothes, and I know its silly, however I believe if the Clothing foundation is strong enough it can keep the immersion somewhat intact since it will hopefully overshadow the sillier harder-to-find things.

## Why It's Good For The Game

This system was created in 2019 for TG and ported last year here with, I believe, little regard for how it would fit or be used.

It's a very limited and unheard-of system that has very little purpose due to its limitations.

You can, in theory, pretend to dress like the HoS, but you can only dye items once with the stamp because it is consumed in the process and you can't do the same for a hat.

Not to mention stamps are hard to get, especially for a single-use item.

This PR breathes new life into this system so that it may be used and become a part of any creative players attempting special gimmicks.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Here you can see the new washing machine sounds, sec dye, jumpskirts actually being dyable and another type of suit being dyable (ignore the suit, now suits have their own Dye Reg and will need to be expanded on in the future, but the example still serves)

https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/1e214197-4e5f-41e8-8f90-2db186056803

Gibbing animals and cleaning, all working as intended.

https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/bf81427d-8156-4290-a6ef-1c7639ac8632

Dyeing and Dyeing removal

https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/ac66ac7a-2901-41fd-9e83-f05b3960796c

ERROR skins

https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/1c992d48-c7e1-40d2-a71d-0723fff2b9bc

Clown Dye, Banana Peel

https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/d8dad2cc-6231-4ba5-9619-b80b7fe4630d

Clown Dye, Banana

https://github.com/BeeStation/BeeStation-Hornet/assets/94647521/0a87b34a-f912-4ce0-a384-228a4c717ddc

Them Berets

image

</details>

Sounds were made by modifying CC0 sounds and one by Nkzdra at https://freesound.org/people/Nkzdra/sounds/347965/

## Changelog
:cl: Solene Yoon
soundadd: Added sounds for the Washing Machine wash cycle.
add: New groups of items are now Dyable (Suits, Hats, Berets and for the clown, Bananas, Practice Laser-Guns and Banana Peels).
tweak: Reworked the REDCOAT dye into DENIED dye of the Denied Stamp.
add: Added ERROR clothing items (Hat, Jumpsuit, Jumpskirt, Shoes and Gloves).
spriteadd: Created Sprites for the ERROR items.
add: DENIED dye will now work on bedsheets to produce what looks like a cultist bedsheet.
remove: Removed the Redcoat Uniform (was spriteless and only used as part of the REDCOAT Dye).
add: Added a White Beret, Rainbow Beret, Mime Invisible Beret and H.O.N.K Tactical Beret.
spriteadd: Created sprites for the Berets above.
fix: Fixed some bugs regarding Washing Machines having to do with Shoes and Jumpskirts being unable to be dyed, Pet Gibbing and Washing Machine cleaning.
tweak: QOL Washing Machine Tweaks: (Wash Cycles clean the Washing Machine, Alt-Clicking will close an open Washing Machine and start a new Wash Cycle).
add: Washing a Dyed item now returns it to normal.
tweak: Accessories (like medals and waistcoats) now retain their colour when attached to a Spray-painted item.
code: added boilerplate code to ensure accessories always appear on top of whatever overlay the clothing may have.
tweak: Colour Sources (Stamps and Crayons) now survive Wash Cycles.
add: Added Space Law book, SOP books and paper as single use colour sources (These don't survive).
tweak: Ensured Colour Dye Registries are attached to the parents of each article of clothing, in order to be able to dye all type of items (e.g you used not to be able to Dye fingerless gloves).
code: added a flippable var for softcaps

/:cl:

### CODE REVIEW

**Comment by BriggsIDP** [comment]
please give them a cool sound 🙏

**Comment by Nakalon** [comment]
> please give them a cool sound 🙏

Wait sound? To the washing machines?...

I hadn't thinked about that... But I may try to do it!

**Comment by Nakalon** [comment]
In my hubris I deleted the only thing making Washing Machines wash things. I am a fool and am reverting this. (I thought it was a signal for cleaning the bloody mess not remembering how it washes CLOTHES)

**Comment by MarkusLarsson421** [comment]
Would it be a good idea to remove the showers ability of instantly cleaning clothes? Would encourage their usage and force those who care of their looks out of their department.

**Comment by Nakalon** [comment]
> Would it be a good idea to remove the showers ability of instantly cleaning clothes? Would encourage their usage and force those who care of their looks out of their department.

That's a genuinely great idea. But then the showers would have no purpose rather than a small mood bonus... I'd need to find some middle ground... But this is a great idea

**Comment by MarkusLarsson421** [comment]
> That's a genuinely great idea. But then the showers would have no purpose rather than a small mood bonus... I'd need to find some middle ground... But this is a great idea

Outside of the dorm showers, most showers are emergency showers, for when you're on fire or whatnot. Not to get clean. 

That mood buff shouldn't happen when you are fully clothed because who the hell showers with their clothes on?!! Make it a mood debuff when you have clothes even.

**Comment by Nakalon** [comment]
> > That's a genuinely great idea. But then the showers would have no purpose rather than a small mood bonus... I'd need to find some middle ground... But this is a great idea
> 
> Outside of the dorm showers, most showers are emergency showers, for when you're on fire or whatnot. Not to get clean.
> 
> That mood buff shouldn't happen when you are fully clothed because who the hell showers with their clothes on?!! Make it a mood debuff when you have clothes even.

while I agree with what you mention, its outside the scope of this PR to change showers and I nee to keep focused!

**Comment by Nakalon** [comment]
Waiting on #10425 to add bandanas

**Comment by Tsar-Salat** [comment]
> ### Here is a list of planned features:
> Color Source is reusable. ✅
> 
> Fix Sneaker Coloring. ✅
> 
> Fix Jumpskirt Coloring. ✅
> 
> Add sound to washing machines. ✅
> 
> Fix Cleaning ✅
> 
> Fix Mob Duplication and Runtime ✅
> 
> Replaced Redcoat Dye with Denied Dye ✅
> 
> * Removed the Redcoat ✅
> * For now this dye is just red dye but it wont be so in the future 📝
> 
> Small QOL changes ✅ Alt-Click close the door if its open and starts the cycle ✅ Cycling a dirty machine cleans it ✅
> 
> Add an Approved Stamp Dye. 📝
> 
> Replaced Redcoat Dye with Denied Dye ✅
> 
> * Removed the Redcoat ✅
> * For now this dye is just red dye but it won't be so in the future 📝
> 
> Make SOP books a color source. ✅
> 
> * Security ✅
> * Space Law ✅
> * Other SOP books are standard colours for now ✅
> 
> Make SOP books only have one use. ✅
> 
> Modify Current dyes to be as different from each other as possible. ✅
> 
> Add Dyes 📝
> 
> * Security Dye ✅
> 
> Make all subtypes of dyable items dyable. ✅
> 
> * Jumpsuits ✅
> * Jumpskirts ✅
> * Gloves ✅
> 
> Expand the items already in the pool of items that can be Dyed ✅
> 
> * Jumpskirts ✅
> 
> Add more items to the pool of items that can be Dyed 📝
> 
> * Hats ✅
> * Berrets ✅
> * Weapons (Practice Laser Gun) ✅
> * Bananas ✅
> * Banana Peels ✅
> 
> Added new colors of:
> 
> * Berets ✅
> * 1 Jumpsuit and Jumpskirt ✅

use the builtin checklist please 

image

Like so:

- [ ] This format hurts the eyes
- [ ] The user has updated the formatting
- [ ] Contributor's eyes are bleeding

**Comment by Nakalon** [comment]
> > ### Here is a list of planned features:
> > Color Source is reusable. ✅
> > Fix Sneaker Coloring. ✅
> > Fix Jumpskirt Coloring. ✅
> > Add sound to washing machines. ✅
> > Fix Cleaning ✅
> > Fix Mob Duplication and Runtime ✅
> > Replaced Redcoat Dye with Denied Dye ✅
> > 
> > * Removed the Redcoat ✅
> > * For now this dye is just red dye but it wont be so in the future 📝
> > 
> > Small QOL changes ✅ Alt-Click close the door if its open and starts the cycle ✅ Cycling a dirty machine cleans it ✅
> > Add an Approved Stamp Dye. 📝
> > Replaced Redcoat Dye with Denied Dye ✅
> > 
> > * Removed the Redcoat ✅
> > * For now this dye is just red dye but it won't be so in the future 📝
> > 
> > Make SOP books a color source. ✅
> > 
> > * Security ✅
> > * Space Law ✅
> > * Other SOP books are standard colours for now ✅
> > 
> > Make SOP books only have one use. ✅
> > Modify Current dyes to be as different from each other as possible. ✅
> > Add Dyes 📝
> > 
> > * Security Dye ✅
> > 
> > Make all subtypes of dyable items dyable. ✅
> > 
> > * Jumpsuits ✅
> > * Jumpskirts ✅
> > * Gloves ✅
> > 
> > Expand the items already in the pool of items that can be Dyed ✅
> > 
> > * Jumpskirts ✅
> > 
> > Add more items to the pool of items that can be Dyed 📝
> > 
> > * Hats ✅
> > * Berrets ✅
> > * Weapons (Practice Laser Gun) ✅
> > * Bananas ✅
> > * Banana Peels ✅
> > 
> > Added new colors of:
> > 
> > * Berets ✅
> > * 1 Jumpsuit and Jumpskirt ✅
> 
> use the builtin checklist please
> 
> image
> 
> Like so:
> 
> * [x]  This format hurts the eyes
> * [ ]  The user has updated the formatting
> * [x]  Contributor's eyes are bleeding
> 
> It formats much better on both mobile and pc web browsers, and reflects in the Repositories PR list

I tried this and I didn't like it.
I'm mainly using this as a personal checklist for the time being, so I rather keep it in way I can navigate easily than a system I don't like and can't navigate that well.

**Comment by Nakalon** [comment]
Since I'm reaching the end I did the thing but now it looks awful

**Comment by Nakalon** [comment]
And I closed it like a dumb dumb

**Comment by Nakalon** [comment]
After speaking to Rkz I decided to atomize at least the clothing part of this PR to get the important and fun changes out in a speedy manner.

I will keep the sprites I made for things like rainbow clown and mime dyed items but remove things like the suits and stuff.

**Comment by Nakalon** [comment]
This bitch is now ready for review. LETS GO!

File: code/game/machinery/washing_machine.dm [hunk 1]
```diff
@@ -117,30 +188,40 @@ GLOBAL_LIST_INIT(dye_registry, list(
...
+
+/obj/machinery/washing_machine/Initialize()
+	. = ..()
+	soundloop = new(src,  FALSE)
 
 /obj/machinery/washing_machine/ComponentInitialize()
 	. = ..()
 	RegisterSignal(src, COMSIG_COMPONENT_CLEAN_ACT, PROC_REF(clean_blood))
 
+/obj/machinery/microwave/Destroy()
```
Comment (HowToLoLu) [inline code/game/machinery/washing_machine.dm]: Did you mean `/obj/machinery/washing_machine/Destroy()`?

File: code/game/machinery/washing_machine.dm [hunk 2]
```diff
@@ -243,6 +361,22 @@ GLOBAL_LIST_INIT(dye_registry, list(
...
+	var/list/dyes = list(DYE_MIME, DYE_CLOWN, DYE_LAW, DYE_CAPTAIN, DYE_HOP, DYE_HOS, DYE_CE)
+	. = ..()
+	if(.)
+		if(dye_color in dyes)
+			flippable = FALSE //Making Hats unflippable so it doesn't break the illusion of the dye if the result isn't a flippable hat
+		else
+			flippable = TRUE
+
+/obj/item/clothing/glasses/machine_wash(obj/machinery/washing_machine/WM)
+	if(HAS_TRAIT(src, TRAIT_SPRAYPAINTED))
```
Comment (HowToLoLu) [inline code/game/machinery/washing_machine.dm]: Should be `HAS_TRAIT_FROM(src, TRAIT_SPRAYPAINTED, CRAYON_TRAIT)` given the code below

File: code/game/machinery/washing_machine.dm [hunk 3]
```diff
@@ -174,10 +255,39 @@ GLOBAL_LIST_INIT(dye_registry, list(
...
+					new_config=initial(item.greyscale_config),
+					new_worn_config=initial(item.greyscale_config_worn),
+					new_inhand_left=initial(item.greyscale_config_inhand_left),
+					new_inhand_right=initial(item.greyscale_config_inhand_right)
+				)
+			else
+				item.icon = initial(item.icon)
+				item.lefthand_file = initial(item.lefthand_file)
+				item.righthand_file = initial(item.righthand_file)
+				item.worn_icon = initial(item.worn_icon)
```
Comment (HowToLoLu) [inline code/game/machinery/washing_machine.dm]: This loop seems redundant if you look into the code above, which utilizes `/atom/movable/proc/machine_wash`, called 3 lines above. which may be the best place to reset the appearance.


A lot of it actually looks like code from the lower half of `/obj/item/proc/dye_item`. Perhaps you could split the functionality of that proc into it and a sort've "copy appearance" function.

File: code/modules/clothing/under/skirt_dress.dm [hunk 4]
```diff
@@ -39,6 +40,7 @@
 	name = "black skirt"
 	desc = "A black skirt, very fancy!"
 	icon_state = "blackskirt"
+	dying_key = DYE_REGISTRY_JUMPSKIRT
```
Comment (HowToLoLu) [inline code/modules/clothing/under/skirt_dress.dm]: Redundant if `/obj/item/clothing/under/dress` is also set to `DYE_REGISTRY_JUMPSKIRT`, unless that was not intended.

File: icons/mob/clothing/head.dmi
Comment (HowToLoLu) [inline icons/mob/clothing/head.dmi:1]: `cowboy_outlaw` seems to've been deleted by accident? Still in use by `/obj/item/clothing/head/helmet/outlaw` in `code\modules\clothing\head\helmet.dm`

image

Comment (Nakalon) [reply icons/mob/clothing/head.dmi:1]: I tried looking for it in my master, still didn't see it... I'll look again...

File: code/modules/clothing/under/skirt_dress.dm (hunk 4 above)
Comment (Nakalon) [reply code/modules/clothing/under/skirt_dress.dm]: Solved!

File: code/game/machinery/washing_machine.dm (hunk 1 above)
Comment (Nakalon) [reply code/game/machinery/washing_machine.dm]: Solved in upcoming commit. I stole this from microwaves :trollface:

File: code/game/machinery/washing_machine.dm (hunk 2 above)
Comment (Nakalon) [reply code/game/machinery/washing_machine.dm]: Crayon trait is always used with spraypainted so they'd be always together, however I added what was missing as it was an oversight

File: icons/mob/clothing/head.dmi
Comment (Nakalon) [reply icons/mob/clothing/head.dmi:1]: Needed to update everything, all looks well

**Comment by Nakalon** [comment]
Dunnu why this is failing...

**Comment by HowToLoLu** [comment]
> Dunnu why this is failing...

If you go to Details and skim the log, you can find unit test failures with a `REASON` message in all caps
"REASON #1: The following clothing items have can_adjust set to true, but have no adjusted icon state: /obj/item/clothing/under/suit/black/female"

File: code/game/machinery/washing_machine.dm [hunk 5]
```diff
@@ -156,24 +237,24 @@ GLOBAL_LIST_INIT(dye_registry, list(
...
 	for(var/X in contents)
 		var/atom/movable/AM = X
 		AM.wash(CLEAN_WASH)
 		AM.machine_wash(src)
-
-	busy = FALSE
+		if(!color_source && isitem(X))
+			var/obj/item/I = X
+			I.appearance_change(I)
+			I.desc = initial(I.desc)
```
Comment (HowToLoLu) [inline code/game/machinery/washing_machine.dm]: Might be better to do this in `appearance_change`

**Comment by Tsar-Salat** [comment]
> Dunnu why this is failing...

I opened a PR on your branch 10 hours ago, just merge that and it should fix it 

https://github.com/Nakalon/BeeStation-Hornet/pull/1

**Comment by Nakalon** [comment]
> > Dunnu why this is failing...
> 
> I opened a PR on your branch 10 hours ago, just merge that and it should fix it
> 
> [Nakalon#1](https://github.com/Nakalon/BeeStation-Hornet/pull/1)

How I didn't notice this is beyond me... Strange...

Thanks for the help!

File: code/game/machinery/washing_machine.dm (hunk 5 above)
Comment (Nakalon) [reply code/game/machinery/washing_machine.dm]: Done!

File: code/game/machinery/washing_machine.dm [hunk 6]
```diff
@@ -286,6 +393,30 @@ GLOBAL_LIST_INIT(dye_registry, list(
...
+					clean_spray.reagents.remove_reagent(/datum/reagent/space_cleaner, clean_spray.amount_per_transfer_from_this,1)
+					playsound(loc, 'sound/effects/spray3.ogg', 50, 1, -6)
+					user.visible_message("[user] has cleaned \the [src].", "<span class='notice'>You clean \the [src].</span>")
+					bloody_mess = 0
+					update_icon()
+				else
+					to_chat(user, "<span class='warning'>You need more space cleaner!</span>")
+				return TRUE
+
+			if(istype(W, /obj/item/soap) || istype(W, /obj/item/reagent_containers/glass/rag))
```
Comment (ike709) [inline code/game/machinery/washing_machine.dm]:
```suggestion
			else if(istype(W, /obj/item/soap) || istype(W, /obj/item/reagent_containers/glass/rag))
```

**Review by ike709 — APPROVED** [review]
Seems largely fine. Where'd you source the sounds from?

**Comment by Nakalon** [comment]
> Seems largely fine. Where'd you source the sounds from?

I created them by editing royalty free sounds!

**Review by BriggsIDP — APPROVED** [review]

**Comment by ike709** [comment]
> I created them by editing royalty free sounds!

Link the original royalty free sounds so we can confirm no attribution is required then ping me on discord for merge.

**Comment by ike709** [comment]
Oh it's CC0: https://freesound.org/people/keweldog/sounds/181148/

**Pull request merged**